### PR TITLE
WIP: Provide more precise reason for cc invalidation

### PR DESCRIPTION
**Pull request by abstratt**
Issue: #26553

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by abstratt** [comment]
@bot-gradle test ACC

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintChecker.kt [hunk 1]
```diff
@@ -310,17 +327,46 @@ class ConfigurationCacheFingerprintChecker(private val host: Host) {
...
+    }
+
+    private
+    fun checkUpToDate(file: File, originalHash: HashCode): UpToDateStatus {
+        val snapshot = host.read(file)
+        if (snapshot.hash == originalHash) {
+            return UpToDateStatus.Unchanged
+        }
+        return when (snapshot.type) {
+            FileType.RegularFile -> if (file.isFile)
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintChecker.kt:355]: 💅 A nested `when` would be clearer

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintChecker.kt [hunk 2]
```diff
@@ -159,8 +161,23 @@ class ConfigurationCacheFingerprintChecker(private val host: Host) {
...
             is ConfigurationCacheFingerprint.InputFile -> input.run {
-                if (hasFileChanged(file, hash)) {
-                    return "file '${displayNameOf(file)}' has changed"
+                val status = checkUpToDate(file, hash)
+                when (status) {
+                    UpToDateStatus.FileChanged -> {
+                        return "file '${displayNameOf(file)}' has changed"
+                    }
+                    UpToDateStatus.Changed -> {
+                        return "file '${displayNameOf(file)}' has somehow changed"
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintChecker.kt:170]: 🤔 somehow? does that add value over simply the "has changed" above?

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheDependencyResolutionFeaturesIntegrationTest.groovy [hunk 3]
```diff
@@ -706,7 +706,7 @@ class ConfigurationCacheDependencyResolutionFeaturesIntegrationTest extends Abst
 
         then:
         configurationCache.assertStateStored()
-        outputContains("Calculating task graph as configuration cache cannot be reused because file '${lockFile}' has changed.")
+        outputContains("Calculating task graph as configuration cache cannot be reused because file '${lockFile}' has been removed.")
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheDependencyResolutionFeaturesIntegrationTest.groovy:709]: 👍

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheDependencyResolutionFeaturesIntegrationTest.groovy [hunk 4]
```diff
@@ -795,7 +795,7 @@ class ConfigurationCacheDependencyResolutionFeaturesIntegrationTest extends Abst
         configurationCacheRun("resolve1")
 
         then:
-        outputContains("Calculating task graph as configuration cache cannot be reused because file 'gradle/verification-metadata.xml' has changed.".replace('/', File.separator))
+        outputContains("Calculating task graph as configuration cache cannot be reused because file 'gradle/verification-metadata.xml' has been removed.".replace('/', File.separator))
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheDependencyResolutionFeaturesIntegrationTest.groovy:798]: 👍

**Comment by abstratt** [comment]
Closing spike PR

**Pull request closed without merging**